Fix whitespace conformance to match UAX31 (including permitting LRM/RLM) (#673)
unicode-org/message-format-wg: spec/syntax.md

### Whitespace

The syntax limits whitespace characters outside of a _pattern_ to the following:
`U+0009 CHARACTER TABULATION` (tab), 
`U+000A LINE FEED` (new line),
`U+000D CARRIAGE RETURN`, 
`U+3000 IDEOGRAPHIC SPACE`, 
or `U+0020 SPACE`.

Inside _patterns_ and _quoted literals_,
whitespace is part of the content and is recorded and stored verbatim.
Whitespace is not significant outside translatable text, except where required by the syntax.

There are two whitespace productions in the syntax.
**_<dfn>Optional whitespace</dfn>_** is whitespace that is not required by the syntax, 
but which users might want to include to increase the readability of a _message_.
**_<dfn>Required whitespace</dfn>_** is whitespace that is required by the syntax.

Tools SHOULD generate `U+200E LEFT-TO-RIGHT MARK` or `U+200F RIGHT-TO-LEFT MARK` 
characters where permitted by the syntax before or following _identifiers_,

> _unquoted literals_, or _option_ values that use right-to-left characters 

Reviewer: I don't see why there is the restriction on **unquoted** literals. Shouldn't it be any literal? That is, anywhere a unquoted literal can appear, and unquoted one can. So it seems like either both (aka just 'literal') or neither can appear.

I could match on:
```
X y ⎨⎨$count⎬⎬
```
Where X is a RTL character, or on
```
⎸X⎸ y ⎨⎨$count⎬⎬
```

In both cases the result is jumbled (disregard the direction of the fake braces, the tool just reorders.
```
⎬	⎬	y		⎨	⎨	$	c	o	u	n	t		X

⎬	⎬	y		⎨	⎨	$	c	o	u	n	t		⎸	X	⎸
```
Now, in this case I could put LRMs before the first ⎸ and after the second, because both positions allow whitespace. Are there circumstances around literals where it makes a difference in the insertability of LRM/RLM because of the quoting?

Reviewer: Other than that, the changes look good to me. Note that if there are any issues in the WG about this, we refrain from these changes until after the v45 release, just leaving a note that we're looking at the bidi ordering issues...

Author: You're right. The key thing I think is to remind tool writers not to quote the mark onto the value.

Author: > Note that if there are any issues in the WG about this, we refrain from these changes until after the v45 release, just leaving a note that we're looking at the bidi ordering issues...

The changes are to the syntax and I think important enough to merit doing the change now--the better to stabilize the syntax. It does represent a relaxation of what is allowed in free whitespace. I would like to avoid having a lot of Tech Preview implementations reject bidi-friendlier messages in the fall. 

OTOH, it does represent a departure from how we set up the `s` production.

Reviewer: Right, for a LTR reading, you want to put LRM around any 'element' that could contain RTL characters. Each literal being matched, literals in option values, etc. So in 

```
{{STUFF {$value option=|JUNK| ...} TO READ}}
```
You want to insert like:
```
{{<LRM>STUFF {<LRM>$value option=<LRM>|JUNK|<LRM> ...<LRM>} TO READ<LRM>}}
```

Of course:
- With LRM/RLM you can't the the RTL message parts to reorder around the {$value}, but the order is predictable and far better than the raw message.
- in practice you want **tooling** to do this, not humans.

Author: > `{{<LRM>STUFF {<LRM>$value option=<LRM>|JUNK|<LRM> ...<LRM>} TO READ<LRM>}}`

You definitely do **not** want the LRMs in the _text_ part of the message, which is where the ones after the `{{` and before the `}}` are. Instead there should be an LRM following the pattern so that any keys in the next variant aren't reordered:

```
<LRM>KEY1<LRM> KEY2<LRM> {{STUFF {<LRM>$value option=<LRM>|JUNK|<LRM>...<LRM>} TO READ}}
<LRM>key1<LRM> key2<LRM> {{ ... next variant...}}
```

Agreed that this is a job for tools.

Reviewer: Whoops, yes, imediately outside {{ and }}, not inside.

### Whitespace

The syntax limits whitespace characters outside of a _pattern_ to the following:
`U+0009 CHARACTER TABULATION` (tab), 
`U+000A LINE FEED` (new line),
`U+000D CARRIAGE RETURN`, 
`U+3000 IDEOGRAPHIC SPACE`, 
or `U+0020 SPACE`.

Inside _patterns_ and _quoted literals_,
whitespace is part of the content and is recorded and stored verbatim.
Whitespace is not significant outside translatable text, except where required by the syntax.

There are two whitespace productions in the syntax.
**_<dfn>Optional whitespace</dfn>_** is whitespace that is not required by the syntax, 
but which users might want to include to increase the readability of a _message_.
**_<dfn>Required whitespace</dfn>_** is whitespace that is required by the syntax.

Tools SHOULD generate `U+200E LEFT-TO-RIGHT MARK` or `U+200F RIGHT-TO-LEFT MARK` 
characters where permitted by the syntax before or following _identifiers_,
_unquoted literals_, or _option_ values that use right-to-left characters 
so that the _message_ displays intelligibly.

These definitions of _whitespace_ implement

> [UTR#31 Rule 3a-2](https://www.unicode.org/reports/tr31/#R3a-2).

Reviewer 2: ```suggestion
[UTR#31 Requirement R3a-2](https://www.unicode.org/reports/tr31/#R3a-2).
```

Reviewer 2: On the classification of UAX, UTS (synchronized and non-synchronized), and UTR, see https://www.unicode.org/reports/about-reports.html.

(And yes, that page does say that a Unicode Technical Report (UTR) is a specific type of Unicode Technical Report. 🤪)

It is a profile of R3a-1 in that specification because:
the following pattern whitespace characters are not allowed:
`U+000B FORM FEED`, 
`U+000C VERTICAL TABULATION`, 
`U+0085 NEXT LINE`, 
`U+2028 LINE SEPARATOR` and 
`U+2029 PARAGRAPH SEPARATOR`;
the character `U+3000 IDEOGRAPHIC SPACE`
_is_ interpreted as whitespace;

>  and the following character is not included in ignorable format controls:

Reviewer 2: ```suggestion
 and the following character is not interpreted as whitespace (in particular, it is not treated as an ignorable format control):
```
This is admittedly pedantry (but what did you expect from a standardizer?), but given the way R3a is written I could technically read « the following character is not included in ignorable format controls » as « that character is whitespace, but not treated as an ignorable format control, and thus is horizontal space ».

Author: Matching suggestion to our formatting.

```suggestion
and the following character is not interpreted as whitespace
(in particular, it is not treated as an ignorable format control):
```

Inside _patterns_ and _quoted literals_,
whitespace is part of the content and is recorded and stored verbatim.
Whitespace is not significant outside translatable text, except where required by the syntax.

There are two whitespace productions in the syntax.
**_<dfn>Optional whitespace</dfn>_** is whitespace that is not required by the syntax, 
but which users might want to include to increase the readability of a _message_.
**_<dfn>Required whitespace</dfn>_** is whitespace that is required by the syntax.

Tools SHOULD generate `U+200E LEFT-TO-RIGHT MARK` or `U+200F RIGHT-TO-LEFT MARK` 
characters where permitted by the syntax before or following _identifiers_,
_unquoted literals_, or _option_ values that use right-to-left characters 
so that the _message_ displays intelligibly.

These definitions of _whitespace_ implement
[UAX#31 Requirement R3a-2](https://www.unicode.org/reports/tr31/#R3a-2).
It is a profile of R3a-1 in that specification because:
the following pattern whitespace characters are not allowed:
`U+000B FORM FEED`, 
`U+000C VERTICAL TABULATION`, 
`U+0085 NEXT LINE`, 
`U+2028 LINE SEPARATOR` and 
`U+2029 PARAGRAPH SEPARATOR`;
the character `U+3000 IDEOGRAPHIC SPACE`

> _is_ interpreted as whitespace;
> and the following character is not interpreted as whitespace
> (in particular, it is not treated as an ignorable format control):
> `U+200F RIGHT-TO-LEFT MARK`. 

Author: Since I added U+200F, this text no longer applies.

```suggestion
_is_ interpreted as whitespace.
```

Reviewer 2: Oh, right.

### Whitespace

**_<dfn>Whitespace</dfn>_** is defined as one or more of
`U+0009 CHARACTER TABULATION` (tab), 
`U+000A LINE FEED` (new line),
`U+000D CARRIAGE RETURN`, 
`U+3000 IDEOGRAPHIC SPACE`, 
or `U+0020 SPACE`,
optionally prepended with `U+200E LEFT-TO-RIGHT MARK`.

Inside _patterns_ and _quoted literals_,
whitespace is part of the content and is recorded and stored verbatim.
Whitespace is not significant outside translatable text, except where required by the syntax.

The character `U+200E LEFT-TO-RIGHT MARK` (LRM) MAY be prepended to _whitespace_ outside
_patterns_ and _quoted literals_ to assist with presentation to users.
Tools SHOULD generate these LRM characters following _identifiers_, _unquoted literals_, or 
_option_ values that use right-to-left characters so that the _message_ displays
intelligibly in a left-to-right context.

This definition of _whitespace_ implements 
[UTR#31 Rule 3a-2](https://www.unicode.org/reports/tr31/#R3a-2).
It is a profile of R3a-1 in that specification because only the

> whitespace characters listed are permitted as whitespace.

Reviewer 2: UAX31-R2a-2 says you need to
> define that profile with a precise specification of the characters that are added to or removed from the set of code points defined by the Pattern_White_Space property, and of any changes to the criteria under which a character or sequence of characters is interpreted as an end of line, as ignorable format controls, or as horizontal space.

The point is that the reader of such a conformance statement sees the difference from the default, which are the things that may need special attention when interoperating with an implementation based on the defaults.

So, if I am reading this right:
- Form feed, next line, line separator, paragraph separator, and right-to-left mark are removed from the set of white space characters;
- ideographic space is added to the set of white space characters;
- the sole ignorable format control (LRM) is allowed only in contexts UAX31-I1 and UAX31-I3 (not UAX31-I2), where UAX31-I1 is further restricted to the beginning of a sequence of horizontal spaces, and UAX31-I3 is further restricted to only the end and not the beginning of a line.

Reviewer 2: > further restricted to only the end and not the beginning of a line

Also it is disallowed on a blank line, I think?

Reviewer 2: > Note that quoted literals or pattern text can contain bidi controls that might cause Trojan source effects unless/until we address placeholder isolation.

Mostly that is something to be dealt with at a higher level (in editors and tooling); the main thing languages need to do is to treat the right things in the right way so that standardized tooling can deal with the issue, see UTS55.

> Note that UAX31-I2 is not permitted because that would break our sigil-identifier syntax.

I don’t really understand what you mean here; UAX31-I2 does not say you should allow `$⟨LRM⟩identifier`, it says that you should allow it _if_ you allow `$    identifier`.

The reason why you do not have UAX31-I2 is that you do not allow an LRM wherever you have `[s]`, so for instance in 

https://github.com/unicode-org/message-format-wg/blob/ae712d70d60e9b5d06bc72d5c7788811aa728f8a/spec/message.abnf#L35-L36

you do not allow `{⟨LRM⟩#whatever⟨LRM⟩}`, only `{⟨LRM⟩ #whatever⟨LRM⟩ }`, right?

Reviewer 2: > you do not allow {⟨LRM⟩#whatever⟨LRM⟩}, only {⟨LRM⟩ #whatever⟨LRM⟩ }, right?

Note that this probably throws a wrench into the UTS55 conversion to plain text algorithm.

Author: Yes. See my response to @gibson042's comment.

`U+0009 CHARACTER TABULATION` (tab), 
`U+000A LINE FEED` (new line),
`U+000D CARRIAGE RETURN`, 
`U+3000 IDEOGRAPHIC SPACE`, 
or `U+0020 SPACE`,
optionally prepended with `U+200E LEFT-TO-RIGHT MARK`.

Inside _patterns_ and _quoted literals_,
whitespace is part of the content and is recorded and stored verbatim.
Whitespace is not significant outside translatable text, except where required by the syntax.

The character `U+200E LEFT-TO-RIGHT MARK` (LRM) MAY be prepended to _whitespace_ outside
_patterns_ and _quoted literals_ to assist with presentation to users.
Tools SHOULD generate these LRM characters following _identifiers_, _unquoted literals_, or 
_option_ values that use right-to-left characters so that the _message_ displays
intelligibly in a left-to-right context.

This definition of _whitespace_ implements 
[UTR#31 Rule 3a-2](https://www.unicode.org/reports/tr31/#R3a-2).
It is a profile of R3a-1 in that specification because:
the following pattern whitespace characters are not allowed:
`U+000B LINE FEED`, 
`U+000C VERTICAL TABULATION`, 
`U+0085 NEXT LINE`, 
`U+2028 LINE SEPARATOR` and 
`U+2029 PARAGRAPH SEPARATOR`;
the character `U+3000 IDEOGRAPHIC SPACE`

> _is_ included in pattern whitespace;

Author: I know, which is why I didn't say Pattern_White_Space. "interpreted as" is somehow not quite right in our space. But we both have other comments below that probably make this line moot.

> _is_ included in pattern whitespace;

Author: ... not to mention that we overload the term "pattern"...

### Whitespace

The syntax limits whitespace characters outside of a _pattern_ to the following:
`U+0009 CHARACTER TABULATION` (tab), 
`U+000A LINE FEED` (new line),
`U+000D CARRIAGE RETURN`, 
`U+3000 IDEOGRAPHIC SPACE`, 
or `U+0020 SPACE`.

Inside _patterns_ and _quoted literals_,
whitespace is part of the content and is recorded and stored verbatim.
Whitespace is not significant outside translatable text, except where required by the syntax.

There are two whitespace productions in the syntax.
**_<dfn>Optional whitespace</dfn>_** is whitespace that is not required by the syntax, 
but which users might want to include to increase the readability of a _message_.
**_<dfn>Required whitespace</dfn>_** is whitespace that is required by the syntax.

> Tools or users SHOULD insert paired isolating controls `U+2066 LEFT-TO-RIGHT ISOLATE`

Reviewer: - The SHOULD should be limited to those pieces that contain RTL text.
- The SHOULD should be limited to "tooling".
I think "or users" sets unrealistic expectations; manual insertion can be tricky, and runs the risk with isolates of having unpaired bidi controls.
Maybe instead add at the end a note that users may also manually insert these characters, 
but tooling can make sure that the bidi controls are in the right positions,
and that all and only the bidi controls necessary are included.
- The two sections about inserting isolates and inserting marked should be clearly separated.

Author: @macchiati 

The SHOULD is already limited to RTL. I added an RTL part of the note about literals.

I split the isolate and marks paragraphs apart.

There are two ways to go on the "tools or users" thing. An alternative to getting into who should do it is to say "Messages SHOULD include..." and "Messages MAY include..." and then having a note about tools, which is how I've handled this.

Note that the syntax requires expression and quoted-pattern controls to be paired (although the whitespace rules break this)

and `U+2069 POP DIRECTIONAL ISOLATE` as permitted by the ABNF _inside_
of _placeholder_ markers `{` and `}` 
and _outside_ _quoted-pattern_ markers `{{` and `}}`,
as well as around _identifiers_,
_literals_ (taking care not to include the mark inside any quotes), 
or _option_ values that use right-to-left characters 
so that the _message_ displays intelligibly.
_Literals_, especially each individual _key_ in a variant, SHOULD be isolated 
using these paired controls.
Tools or users MAY also insert `U+200E LEFT-TO-RIGHT MARK` or `U+200F RIGHT-TO-LEFT MARK`
characters where permitted by the syntax before or following _identifiers_,
_literals_ (taking care not to include the mark inside any quotes), 
or _option_ values that use right-to-left characters 
so that the _message_ displays intelligibly.

These definitions of _whitespace_ implement
[UAX#31 Requirement R3a-2](https://www.unicode.org/reports/tr31/#R3a-2).
It is a profile of R3a-1 in that specification because:
the following pattern whitespace characters are not allowed:
`U+000B FORM FEED`, 
`U+000C VERTICAL TABULATION`, 
`U+0085 NEXT LINE`, 
`U+2028 LINE SEPARATOR` and 
`U+2029 PARAGRAPH SEPARATOR`;

> the character `U+3000 IDEOGRAPHIC SPACE`
> _is_ interpreted as whitespace.

Reviewer 2: ```suggestion
the character `U+3000 IDEOGRAPHIC SPACE`
_is_ interpreted as whitespace,
as well as the directional isolates U+2066..U+2069
(these having the Default_Ignorable_Code_Point property, they are ignorable format controls).
```

Author: Is that true, though? I added them directly to `owsp`, meaning they can be used by themselves in optional whitespace locations (without having to insert any actual whitespace), while `wsp` requires that these characters (optionally) appear outside (before/after) a required run of whitespace. We don't impute any "whitespace-ness" to the controls. They're just doing their default-ignorable jobs 😉

Reviewer 2: Yes, but being an ignorable format control (which does not count as a required space, and is only allowed as an optional space) counts as one of the three ways in which something can be _interpreted as whitespace_ in the sense of [UAX31-R3a](https://www.unicode.org/reports/tr31/#R3a) (the other two being horizontal space, and line terminator or part of a line terminator).

Reviewer 2: (Yes, this is terribly confusing, and part of the general pattern that in Unicode contexts _whitespace_ means a host of different things, from [characters that are whitespace](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3Awhitespace%3A%5D&g=&i=) to [characters that should be interpreted as whitespace in parsing](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3Apattern_white_space%3A%5D&g=&i=) to [characters that are categorized as spaces](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3Aspace+separator%3A%5D&g=&i=) to the zero width space and no-break space that are called spaces, but are not spaces at all and instead are controls serving as overrides for some segmentation algorithms.

And then of course you have [characters that behave like space in line breaking](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3Aline_break%3Dspace%3A%5D&g=&i=), [characters that behave like space in bidirectional ordering](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3Abc%3Dwhite_space%3A%5D&g=&i=), etc.

Unicode Properties & Algorithms are an interesting place.)

Author: Yes, I'm aware. Note well that these controls are not in the `s` production, which is what we call whitespace. I realize that other interpretations of whitespace are possible in Unicode. Your note addition would, I think, be confusing, since the controls are optional and are definitely *not* counted as whitespace in required whitespace parts of our syntax.

Is there harm in leaving this out? If so, I'll commit it.

Reviewer 2: I think it the conformance statement really ought to mention that you are adding to the set of ignorable format controls, but I agree that my proposed wording is needlessly pedantic and confusing. How about this:
```suggestion
the character `U+3000 IDEOGRAPHIC SPACE`
_is_ interpreted as whitespace,
 and the directional isolates U+2066..U+2069
 are treated as ignorable format controls.
```

Author: Works for me! Done.

> [!NOTE]
> The character U+3000 IDEOGRAPHIC SPACE is included in whitespace for
> compatibility with certain East Asian keyboards and input methods,
> in which users might accidentally create these characters in a _message_.

```abnf

> ; optional whitespace
> owsp = *( s / %x200E / %x200F )
> 
> ; required whitespace
> wsp = [ (%x200E / %x200F) ] 1*s [ (%x200E / %x200F) ]
> 
> ; whitespace characters
> s = ( SP / HTAB / CR / LF / %x3000 )

Author: Update to match:

```suggestion
; optional whitespace
owsp = *( s / %x200E / %x200F / %x2066-2069 )

; required whitespace
wsp = [ (%x200E / %x200F / %x2066-2069 ) ] 1*s [ (%x200E / %x200F / %x2066-2069 ) ]

; whitespace characters
s = ( SP / HTAB / CR / LF / %x3000 )
```

**_<dfn>Optional whitespace</dfn>_** is whitespace that is not required by the syntax, 
but which users might want to include to increase the readability of a _message_.
**_<dfn>Required whitespace</dfn>_** is whitespace that is required by the syntax.

_Messages_ SHOULD include paired isolating controls `U+2066 LEFT-TO-RIGHT ISOLATE`
and `U+2069 POP DIRECTIONAL ISOLATE` as permitted by the ABNF
_inside_ of _placeholder_ markers `{` and `}` 
and _outside_ _quoted-pattern_ markers `{{` and `}}`
if either of these contain right-to-left characters,
as well as around _identifiers_,
_literals_ (taking care not to include the mark inside any quotes), 
or _option_ values that use right-to-left characters,
so that the _message_ displays intelligibly.
_Literals_, especially each individual _key_ in a variant, that use right-to-left
characters SHOULD be isolated using these paired controls.

_Messages_ MAY also include `U+200E LEFT-TO-RIGHT MARK` or `U+200F RIGHT-TO-LEFT MARK`
characters where permitted by the syntax before or following _identifiers_,
_literals_ (taking care not to include the mark inside any quotes), 
or _option_ values that use right-to-left characters 

> so that the _message_ displays intelligibly.

Author: The guidance on "never inside" is not necessarily correct, unless one adds something like "never inside unless intended to be part of the message's output"

I will add merge your suggestion, and then will edit it for spec styling requirements (and grammar)

but which users might want to include to increase the readability of a _message_.
**_<dfn>Required whitespace</dfn>_** is whitespace that is required by the syntax.

_Messages_ that contain right-to-left (aka RTL) characters SHOULD use one of the 
following mechanisms to make messages display intelligibly in plain-text editors:

1. Use paired isolating bidi controls `U+2066 LEFT-TO-RIGHT ISOLATE`
   and `U+2069 POP DIRECTIONAL ISOLATE` as permitted by the ABNF around
   parts of any _message_ containing RTL characters:
   - _inside_ of _placeholder_ markers `{` and `}` 
   - _outside_ _quoted-pattern_ markers `{{` and `}}`
   - _identifiers_
   - _literals_ (This is especially important for individual _keys_ in a _variant_)
   - _option_ values
2. Use the 'local-effect' bidi controls`U+200E LEFT-TO-RIGHT MARK` or
   `U+200F RIGHT-TO-LEFT MARK` as permitted by the ABNF around
   parts of any _message_ containing RTL characters:
   -  _identifiers_
   - _literals_ (taking care not to include the mark inside any quotes), 
   - _option_ values

> [!IMPORTANT]
> Always take care **not** to add a bidi control where it is semantically significant:

> > - put them outside of _literal_ quotes, such as `<LRM>|...|<LRM>`
> > - put them outside of quoted _patterns_, such as `<LRI>{{...}}<PDI>`

Reviewer 3: I suggest adding words Just to make it clear what the list is telling people to _not_ do (vs. whether they are supposed to do it). Also, I think you want a newline to ensure that the following sentences of the note don't get collapsed into the second bullet point.

```suggestion
> - do not put them outside of _literal_ quotes, such as `<LRM>|...|<LRM>`
> - do not put them outside of quoted _patterns_, such as `<LRI>{{...}}<PDI>`
>
```

Author: Good point. I am making changes to this text as I incorporate it into a new PR that replaces this one and implements the bidi design.
